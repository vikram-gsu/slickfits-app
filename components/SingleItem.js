import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import Head from 'next/head';
import ErrorMessage from './ErrorMessage';

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

const StyledItem = styled.div`
  max-width: 1200px;
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin-left: 2rem;
    text-align: left;
    padding-top: 1rem;
  }
`;

const SingleItem = ({ id }) => {
  return (
    <Query query={SINGLE_ITEM_QUERY} variables={{ id }}>
      {({ data, loading, error }) => {
        if (error) return <ErrorMessage errorObj={error} />;
        if (loading) return <div>Loading...</div>;
        if (!data.item) return <p>No Item Found for {id}</p>;
        const { title, description, largeImage } = data.item;
        return (
          <StyledItem>
            <Head>
                <title>Slick Fits | {title}</title>
            </Head>
            <img src={largeImage} alt={title} />
            <div className="details">
            <h2>{title}</h2>
            <p>{description}</p>
            </div>
          </StyledItem>
        );
      }}
    </Query>
  );
};

export default SingleItem;
export {SINGLE_ITEM_QUERY};