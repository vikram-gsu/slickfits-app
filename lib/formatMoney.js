function formatMoney(amount) {
  var options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  };
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const intlFormatter = new Intl.NumberFormat('en-US', options);
  return intlFormatter.format(amount / 100);
}
export default formatMoney;
