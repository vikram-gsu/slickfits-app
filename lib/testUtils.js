/* eslint-disable import/no-extraneous-dependencies */
import casual from 'casual';

casual.seed(777);

export const fakeItem = () => ({
    __typename: 'Item',
    id: '12321jlk', 
    title: 'Adidas shoes', 
    description: 'Yellow shoes with three stripes!', 
    price: 3450, 
    image: "adidas.jpg", 
    largeImage: "large_adidas.jpg" 
});

export const fakeUser = () => ({
    __typename: 'User',
    id: '23212',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN']
})