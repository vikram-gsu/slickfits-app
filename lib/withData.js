import withApollo from 'next-with-apollo';
import { ApolloClient } from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';
import { InMemoryCache } from 'apollo-cache-inmemory';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000'
});

function createClient({ headers }) {
  return new ApolloClient({
    link: httpLink,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    cache: new InMemoryCache()
  });
}

export default withApollo(createClient);
