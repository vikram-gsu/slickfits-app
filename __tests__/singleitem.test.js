import { shallow, mount } from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json';
import {GraphQLError} from 'apollo-boost';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';
import { fakeItem } from '../lib/testUtils';

describe('<SingleItem />', () => {
  it('renders', () => {
    shallow(<SingleItem id="123" />);
  });

  it('renders to a snapshot', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: { data: { item: fakeItem() } }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');
    // await wait(0);
    await new Promise(resolve => setTimeout(resolve, 0));
    wrapper.update();
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });

  it('Errors with a not found item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: {
        //   error: [{ message: 'Items Not Found!' }],
        //   data: []
        error: new Error('Item not found!')
        // errors: [new GraphQLError('Item not found!')]
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SingleItem id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    // console.log(item.debug());
    expect(item.text()).toContain('Error');
    expect(toJSON(item)).toMatchSnapshot();
  });

});
