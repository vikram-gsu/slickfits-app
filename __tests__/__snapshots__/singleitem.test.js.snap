// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> Errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Oops!
  </strong>
  Network error: Error writing result to store for query:
 {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"SINGLE_ITEM_QUERY"},"variableDefinitions":[{"kind":"VariableDefinition","variable":{"kind":"Variable","name":{"kind":"Name","value":"id"}},"type":{"kind":"NonNullType","type":{"kind":"NamedType","name":{"kind":"Name","value":"ID"}}},"directives":[]}],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"item"},"arguments":[{"kind":"Argument","name":{"kind":"Name","value":"where"},"value":{"kind":"ObjectValue","fields":[{"kind":"ObjectField","name":{"kind":"Name","value":"id"},"value":{"kind":"Variable","name":{"kind":"Name","value":"id"}}}]}}],"directives":[],"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"title"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"description"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"price"},"arguments":[],"directives":[]},{"kind":"Field","name":{"kind":"Name","value":"largeImage"},"arguments":[],"directives":[]}]}}]}}],"loc":{"start":0,"end":148}}
Cannot read property 'item' of undefined
</p>
`;

exports[`<SingleItem /> renders to a snapshot 1`] = `
<h2>
  Adidas shoes
</h2>
`;

exports[`<SingleItem /> renders to a snapshot 2`] = `
<img
  alt="Adidas shoes"
  src="large_adidas.jpg"
/>
`;

exports[`<SingleItem /> renders to a snapshot 3`] = `
<p>
  Yellow shoes with three stripes!
</p>
`;
